src/dmd/nogc.d

```diff
@@ -81,6 +81,17 @@ public:
             }
             f.printGCUsage(e.loc, "setting `length` may cause a GC allocation");
         }
+        else if (fd.ident == Id._d_arrayappendT || fd.ident == Id._d_arrayappendcTX)
+        {
+            if (f.setGC())
+            {
+                e.error("cannot use operator `~=` in `@nogc` %s `%s`",
+                    f.kind(), f.toPrettyChars());
+                err = true;
+                return;
+            }
+            f.printGCUsage(e.loc, "operator `~=` may cause a GC allocation");
+        }
```

Comment:
I don't quite understand the need for these error messages.  If the lowered expression causes a GC allocation, then the error message will already be emitted there.  However, the error message will be in terms of the lowered code, and not the user code.

If the purpose of this is to make the error message more relevant to user code, then that is the case that I mentioned earlier on slack.  What we could do instead is pass the file, line number, and even the user's code to the runtime hook and have the runtime hook emit a more meaningful error message, perhaps with a static assert.

Regardless, I'm ok with keeping this as is to maintain the status quo.

Reply:
Yes, long-term moving error generation into the hooks would be nicer*. For now I think this is fine to keep compatibility for the error messages.

(*) However, it would also somewhat complicate the 'GC use in ctfe only code is OK' approach: Right now, `main1` and `main2` won't compile:

```d
string aCTFEOnlyHelper()(string a)
{
    return "Hello" ~ a;
}

void nogcAppend(ref string r, string val) @nogc
{
}

void main1() @nogc
{
    string result;
    __ctfe ? result ~= "foo" : nogcAppend(result, "foo");
}

void main2() @nogc
{
    enum val = aCTFEOnlyHelper("foo");
}
```

With the changes suggested [here](https://github.com/dlang/dmd/pull/9982/commits/51927b50211460b97f9a3b7c576a363de9e300fb#r310355871), `main1` will actually compile.

For main2, you have the problem that as soon as you instantiate  aCTFEOnlyHelper, even if it's in a CTFE-only context, the code for aCTFEOnlyHelper will also be emitted to the object file, so you need the GC there. This is probably harder to fix than the `main1` problem, I've filed an issue here: https://issues.dlang.org/show_bug.cgi?id=20101

(Actually, we writing the issue I had epiphany: What we want there is actually already largely available in `speculative template instantiation`. So it might actually not be too difficult to get the main2 case working).

Reply:
This code was moved from `visit(CatAssignExp)` as a call to `_d_arrayappend{T,cTX}` is now what symbolizes array appending, and the message will be printed when compiling with `-vgc`.